### PR TITLE
Make Cloud Logging writes asynchronous via CloudLoggingHandler

### PR DESCRIPTION
**Pull request by findmyway**
`_CloudLogger.write_cloud_logging_entry` previously called `Logger.log_struct`
synchronously, blocking the calling thread on a ~50-150 ms gRPC RTT per
call. On the per-step critical path of a training loop (`record_step_*_time`,
`record_data_loading_*_time`, ...), this added up to non-trivial wall-time
overhead per step.

This PR makes writes asynchronous by delegating to google-cloud-logging's
`CloudLoggingHandler`, whose default transport is `BackgroundThreadTransport`.
Rather than rolling our own daemon-thread + queue + drain loop, we reuse
the upstream worker that already handles atexit, graceful shutdown, batched
commit, and error logging.

> **Note**: this PR has been substantially rewritten from the original
> opt-in `enable_background_writes` design after reviewer feedback. The
> implementation went from ~150 LoC to ~30 LoC by delegating to
> `BackgroundThreadTransport` instead of reimplementing it.

## Public surface

- `_CloudLogger(..., background_grace_period_s=5.0, background_batch_size=100, background_max_latency_s=2.0)`
- `GoodputRecorder` forwards all three tuning knobs.
- `_CloudLogger.flush()`: drain the background queue (= `handler.flush()` → `queue.join()`).

## Why delegate to `CloudLoggingHandler`

- **N entries committed in 1 `write_entries` RPC** (vs N individual `log_struct` calls).
- **LogRecord.created is captured at the `.info()` call site**, so the GCP entry timestamp reflects when the caller wrote — `GoodputCalculator`'s `timestamp>"…"` filtering remains accurate even when the worker flushes later.
- **`BackgroundThreadTransport.grace_period`** bounds shutdown if the GCP endpoint is slow; no atexit hang risk.
- **`_safely_commit_batch`** logs commit errors via `google.cloud.logging`'s logger — no silent drop.

## BREAKING CHANGES (suggest bumping to 0.1.0)

- Callers that read back via `GoodputCalculator` immediately after a write must now call `_CloudLogger.flush()` first.
- Callers that relied on synchronous error propagation from `log_struct` will now see errors via `google.cloud.logging`'s logger instead.

## Tests

- `test_write_logs_entry` rewritten to assert the async route via `_async_handler.emit` (capturing the `LogRecord` with the dict payload).
- 3 new tests in `CloudLoggerAsyncWritesTest`: handler attachment with `propagate=False`, no-op for foreign-job entries, `flush()` forwards to handler.

## Reviewer-comment status

- @Ethanlm "default should be True": resolved — async is now the only path, no flag.
- @YixuanWang-99 "drain timing issue": resolved — upstream `_get_many` doesn't have this bug (drains on every wake until empty or batch_size).
- @dipannita08 "make knobs configurable through the Recorder": resolved — `background_grace_period_s` / `background_batch_size` / `background_max_latency_s` are all forwarded.
- @dipannita08 "profile traces in description": pending — will collect in a follow-up before merging.

## TODO before merge

- [x] Profile traces (xprof / per-step host timeline before & after) on a 7x-256  AXLearn run.

Before: 

Screenshot 2026-06-07 at 21 23 08

Screenshot 2026-06-07 at 21 08 38

After:

Screenshot 2026-06-07 at 21 23 29

Screenshot 2026-06-07 at 21 10 26

- [x] CLA signature.

### CODE REVIEW

File: ml_goodput_measurement/src/goodput.py [hunk 1]
```diff
@@ -78,6 +80,10 @@ def __init__(
       job_name: str,
       log_name: str,
       max_logs_retention_period: Optional[datetime.timedelta] = None,
+      *,
+      enable_background_writes: bool = False,
```
Comment (Ethanlm) [inline ml_goodput_measurement/src/goodput.py]: Maybe it should just default to True. Having the goodput recorder slowing down the critical training path on every step is not a good default

File: ml_goodput_measurement/src/goodput.py [hunk 2]
```diff
@@ -110,12 +161,70 @@ def write_cloud_logging_entry(self, entry) -> None:
...
       self.logger.log_struct(
           entry,
           severity='INFO',
       )
 
+  def _writer_loop(self) -> None:
+    """Daemon worker: periodically drain `_write_queue` to Cloud Logging."""
+    assert self._write_queue is not None and self._writer_stop is not None
+    while not self._writer_stop.is_set():
+      # Wait blocks until the interval elapses OR `flush()` sets the event.
```
Comment (YixuanWang-99) [inline ml_goodput_measurement/src/goodput.py]: Should we drain_once() before waiter_stop()? Now this design can only drain up to background_max_batch_size * background_flush_interval_s

File: ml_goodput_measurement/src/goodput.py (hunk 1 above)
Comment (dipannita08) [reply ml_goodput_measurement/src/goodput.py]: Thanks! Could you add a profile trace for both true and false in the PR description?

@YixuanWang-99, can you do quick test on MaxText as well to get more confidence on default True?

File: ml_goodput_measurement/src/goodput.py [hunk 3]
```diff
@@ -78,6 +80,10 @@ def __init__(
       job_name: str,
       log_name: str,
       max_logs_retention_period: Optional[datetime.timedelta] = None,
+      *,
+      enable_background_writes: bool = False,
+      background_flush_interval_s: float = 2.0,
```
Comment (dipannita08) [inline ml_goodput_measurement/src/goodput.py]: Let's make these configurable through the Recorder base class. Looks like we are hardcoding this everywhere at the moment.

File: ml_goodput_measurement/src/goodput.py [hunk 4]
```diff
@@ -110,12 +161,70 @@ def write_cloud_logging_entry(self, entry) -> None:
...
+  def _writer_loop(self) -> None:
+    """Daemon worker: periodically drain `_write_queue` to Cloud Logging."""
+    assert self._write_queue is not None and self._writer_stop is not None
+    while not self._writer_stop.is_set():
+      # Wait blocks until the interval elapses OR `flush()` sets the event.
+      self._writer_stop.wait(self._flush_interval_s)
+      self._drain_once()
+    # Final pass once the stop signal is observed, to flush whatever
+    # arrived after the last interval but before shutdown.
+    self._drain_once()
```
Comment (YixuanWang-99) [inline ml_goodput_measurement/src/goodput.py]: This only flush up to background_max_batch_size, is it potentially losing logs (logging more than background_max_batch_size)?

**Review by dipannita08 — COMMENTED** [review]
Thank you @findmyway! Did a first pass, I'll be on the look out for your update.

File: ml_goodput_measurement/src/goodput.py (hunk 1 above)
Comment (findmyway) [reply ml_goodput_measurement/src/goodput.py]: @Ethanlm changed to async by default.

**Comment by YixuanWang-99** [comment]
Patched the ml_goodput_measurement library including changes of this PR, Tested on TPU v6e with synthetic data and steps=20:

`python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml base_output_directory=/dev/shm/goodput dataset_type=synthetic run_name=SYNTHETIC_TEST enable_goodput_recording=True steps=20 monitor_goodput=True`

The final logs are:
`I0608 20:54:47.917440 123567287061632 checkpointing.py:1007] Waiting for step 19 to finish before checkpoint...
I0608 20:54:47.918323 123567287061632 checkpointing.py:1011] Waited 0.0009036064147949219 seconds for step 19 to finish before starting checkpointing.
I0608 20:54:47.918582 123567287061632 checkpoint_manager.py:2009] [process=0][thread=MainThread][wait_until_finished] No Save Finalize thread to wait for. Returning.
I0608 20:54:47.918725 123567287061632 checkpoint_manager.py:1512] [process=0] Saving checkpoint at step 19
I0608 20:54:47.918797 123567287061632 event_tracking.py:70] [process=0] [async] Started save checkpoint @ /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.
I0608 20:54:47.924288 123478013380160 atomicity.py:140] Creating tmp directory /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp
I0608 20:54:47.931061 123567287061632 jax_array_handlers.py:378] Scheduling D2H of 39 prioritized jax.Array.
I0608 20:54:47.931213 123567287061632 replica_slices.py:425] Transferring arrays to host memory with options: use_replica_parallel=True, min_slice_bytes_for_replica_parallel=None, max_replicas_for_replica_parallel=None, enable_pinned_host_transfer=False
I0608 20:54:47.932387 123478055323200 atomicity.py:140] Creating tmp directory /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp/items.orbax-checkpoint-tmp
I0608 20:54:48.102377 123567287061632 base_pytree_checkpoint_handler.py:156] [process=0][thread=MainThread] Initiated "orbax.checkpoint._src.serialization.jax_array_handlers.ArrayHandler".serialize. Time taken: 0.172256s
I0608 20:54:48.102552 123567287061632 base_pytree_checkpoint_handler.py:764] MemoryRegulated: Peak usage: 0.000000 GiB
I0608 20:54:48.103251 123567287061632 base_pytree_checkpoint_handler.py:132] [process=0] /jax/orbax/write/blocking_gbytes_per_sec: 69.704 GiB/s (total gbytes: 12.3 GiB) (time elapsed: 0.17704272270202637 s) (per-host)
I0608 20:54:48.103328 123567287061632 base_pytree_checkpoint_handler.py:814] [process=0][thread=MainThread] Initiated Pytree async_save. Time taken: 0.177137s (batch_requests_ready=0.001512s, total_serialization_initiated=0.174900s, others=0.000725s)
I0608 20:54:48.103402 123567287061632 composite_checkpoint_handler.py:715] [process=0][thread=MainThread] Initiated CompositeCheckpointHandler.async_save. Time taken: 0.178958s (all_items=0.000018s, per_item={'items': '0.00001788'}, temp_paths=0.178940)
I0608 20:54:48.114468 123567287061632 event_tracking.py:125] [process=0] [async] Finished blocking save in 0.20 seconds. Continuing save @ /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.
I0608 20:54:48.114907 123478080489024 async_checkpointer.py:76] [process=0][thread=async_save] Background save thread started. Deadline for this save operation is 2026-06-08 21:14:48.114874
I0608 20:54:48.115847 123567287061632 checkpoint_manager.py:1560] [process=0][thread=MainThread][step=19] Starting CheckpointManager Save Finalize thread=save_finalize
I0608 20:54:48.120761 123478023865920 async_checkpointer.py:280] [process=0][thread=save_finalize] Waiting for background save thread=async_save.
I0608 20:54:48.120880 123567287061632 standard_logger.py:34] {'step': 19, 'event_type': 'save', 'directory': '/dev/shm/goodput/SYNTHETIC_TEST/checkpoints', 'reached_preemption': False, 'preemption_received_at': None, 'synchronous': False, 'wait_for_prev_start_time': 1780952087.918556, 'wait_for_prev_duration_secs': 4.9114227294921875e-05, 'time_between_consecutive_saves_sec': 12.005356311798096, 'checkpointer_blocking_start_time': 1780952087.918747, 'checkpointer_blocking_duration_secs': 0.19700980186462402, 'get_old_steps_start_time': 1780952088.1157749, 'get_old_steps_duration_secs': 2.5033950805664062e-05, 'checkpoint_manager_blocking_start_time': 1780952087.9185174, 'checkpoint_manager_blocking_duration_secs': 0.20233583450317383}
I0608 20:54:48.122836 123567287061632 checkpointing.py:563] Started an asynchronous checkpoint save for step 19
I0608 20:54:48.122986 123567287061632 checkpoint_manager.py:2020] [process=0][thread=MainThread][step=19][wait_until_finished] Waiting for Save Finalize thread (save_finalize) to complete.
I0608 20:54:48.144568 123478055323200 array_metadata_store.py:203] [process=0][thread=array_type_handler] Wrote 39 array_metadata.ArrayMetadata to /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp/items.orbax-checkpoint-tmp/array_metadatas/process_0
I0608 20:54:50.000835 123478034351680 base_pytree_checkpoint_handler.py:1328] [process=0][thread=write_metadata_after_commits] Commit + Array metadata written. Time taken: 1.887419s (commit=1.885207s, array_metadata_write=0.002212s)
I0608 20:54:50.001850 123478080489024 base_pytree_checkpoint_handler.py:132] [process=0] /jax/orbax/write/gbytes_per_sec: 5.945 GiB/s (total gbytes: 12.3 GiB) (time elapsed: 2.0756208896636963 s) (per-host)
I0608 20:54:50.001960 123478080489024 async_checkpointer.py:90] [process=0][thread=async_save] 3 Handler Commit operations completed. Time taken: 1.884768s.
I0608 20:54:50.004434 123478080489024 array_metadata_store.py:367] [process=0][thread=async_save] Skipped cross-host ArrayMetadata validation because only one process is found: process_index=0.
I0608 20:54:50.004581 123478080489024 ocdbt_utils.py:49] Param validation support for Zarr3 will be added later (b/362328389).
I0608 20:54:50.005908 123478080489024 base_pytree_checkpoint_handler.py:1452] [process=0][thread=async_save] Pytree save finalize (merge_ocdbt + ArrayMetadata validation) completed. Time taken: 0.003208s. use_zarr3=True, enable_post_merge_validation=True, directory=/dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp/items.orbax-checkpoint-tmp
I0608 20:54:50.006434 123478080489024 atomicity.py:535] Renaming /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp/items.orbax-checkpoint-tmp to /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp/items
I0608 20:54:50.007147 123478080489024 atomicity.py:535] Renaming /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19.orbax-checkpoint-tmp to /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19
I0608 20:54:50.008083 123478080489024 atomicity.py:847] [process=0][thread=async_save] Finished saving checkpoint (finalized tmp dir) to `/dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19`.
I0608 20:54:50.008396 123478080489024 event_tracking.py:138] [process=0] [async] Finished save (blocking + background) in 2.09 seconds @ /dev/shm/goodput/SYNTHETIC_TEST/checkpoints/19
I0608 20:54:50.009155 123478080489024 async_checkpointer.py:160] [process=0][thread=async_save] Background save thread done. Time taken: 1.891963s.
I0608 20:54:50.009270 123478023865920 async_checkpointer.py:288] [process=0][thread=save_finalize] Done with waiting for background save thread=async_save.
I0608 20:54:50.009319 123478023865920 async_checkpointer.py:298] [process=0][thread=save_finalize] No errors found in background save thread=async_save.
I0608 20:54:50.009375 123478023865920 checkpoint_manager.py:2137] [process=0][thread=save_finalize][step=19] CheckpointManager Save Finalize is syncing with other hosts...
I0608 20:54:50.009418 123478023865920 checkpoint_manager.py:2146] [process=0][thread=save_finalize][step=19] CheckpointManager Save Finalize is done on all hosts.
I0608 20:54:50.009711 123567287061632 checkpoint_manager.py:2032] [process=0][thread=MainThread][step=19][wait_until_finished] Done waiting for Save Finalize thread (save_finalize) running at step=19.
I0608 20:54:50.010076 123567287061632 checkpoint_manager.py:2009] [process=0][thread=MainThread][wait_until_finished] No Save Finalize thread to wait for. Returning.
I0608 20:54:50.011064 123567287061632 metric_logger.py:207] completed step: 19, seconds: 0.762, TFLOP/s/device: 213.888, Tokens/s/device: 32239.572, total_weights: 196608, loss: 7.551, lm_loss: 7.551, perplexity: 1902.090
I0608 20:54:50.050677 123567287061632 monitoring.py:767] Shutting down cumulative goodput process (PID: 953806)
W0608 20:55:00.060125 123567287061632 monitoring.py:771] Cumulative goodput process (PID: 953806) did not exit gracefully. Terminating forcefully.
W0608 20:55:34.661050 123567287061632 goodput.py:1341] Unproductive time for BadputType.DATA_LOADING_ASYNC could not be computed.
I0608 20:55:34.802749 123567287061632 gcp_metrics.py:123] {'message': 'Sent Goodput metrics to GCM Monitoring.', 'metrics_count': 14, 'job_name': 'SYNTHETIC_TEST', 'pid': 953806, 'worker': 'Process-1', 'metrics_type': 'cumulative'}
I0608 20:55:34.803411 123567287061632 monitoring.py:104] [PID: 953806] Goodput worker process for job SYNTHETIC_TEST stopped.
W0608 20:56:07.987422 123567287061632 goodput.py:1341] Unproductive time for BadputType.DATA_LOADING_ASYNC could not be computed.
I0608 20:56:08.115210 123567287061632 gcp_metrics.py:123] {'message': 'Sent Goodput metrics to GCM Monitoring.', 'metrics_count': 14, 'job_name': 'SYNTHETIC_TEST', 'pid': 952169, 'worker': 'MainProcess', 'metrics_type': 'cumulative'}
I0608 20:56:08.115353 123567287061632 monitoring.py:817] Final goodput query and upload for job: SYNTHETIC_TEST and logger: goodput_SYNTHETIC_TEST completed with total goodput: 70.68%, last step: 19
I0608 20:56:08.115440 123567287061632 goodput.py:87] Flushed final metrics and safe exited from Goodput monitoring.`

I did not see any hanging process, the metric monitoring is shut down gracefully.

**Comment by YixuanWang-99** [comment]
pytest passed as well:

pytest tests/unit/goodput_utils_test.py
================================================ test session starts =================================================
platform linux -- Python 3.12.13, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/yixuannwang_google_com/projects/maxtext
configfile: pytest.ini
plugins: mock-3.15.1, hypothesis-6.142.1, typeguard-2.13.3, xdist-3.8.0, anyio-4.13.0, jaxtyping-0.3.9
collected 5 items                                                                                                    

tests/unit/goodput_utils_test.py .....                                                                         [100%]

================================================= 5 passed in 14.51s =================================================

File: ml_goodput_measurement/src/goodput.py [hunk 5]
```diff
@@ -102,19 +124,59 @@ def __init__(
...
-          severity='INFO',
-      )
+    if entry[_JOB_NAME] != self.job_name:
+      return
+    # CloudLoggingHandler routes dict msg → StructEntry == log_struct
+    # semantics. LogRecord.created is set at the .info() call site, so
+    # the eventual GCP entry timestamp reflects when the caller wrote.
+    self._async_logger.info(entry)
+
+  def flush(self) -> None:
```
Comment (dipannita08) [inline ml_goodput_measurement/src/goodput.py:165]: _CloudLogger should stay internal. Callers should have an API to flush through the GoodputRecorder. Can you expose a simple public flush() method on GoodputRecorder that calls _cloud_logger's flush under the hood?

File: CHANGELOG.md [hunk 6]
```diff
@@ -20,6 +20,32 @@ To release a new version (e.g. from `1.0.0` -> `2.0.0`):
   * Update the `[Unreleased]` url: `v1.0.0...HEAD` -> `v2.0.0...HEAD`
 
 -->
+## [Unreleased]
+
+### Changed (BREAKING)
+
+* `_CloudLogger.write_cloud_logging_entry` is now asynchronous. Writes are
```
Comment (dipannita08) [inline CHANGELOG.md]: I was wondering if we could make this non-breaking by default so existing callers don't need to update their code. And I think its doable.

Applications expect to write to and read from the same logger bucket for the workload whether they instantiate a GoodputCalculator inline or run a continuous background monitor process. Requiring top-level application code to manually orchestrate .flush() calls may not be required.. and also forces a breaking change.

Could we encapsulate the flushing/synchronization logic directly within the read path (e.g., having GoodputCalculator or the logger coordinate pending writes before querying GCP)? If reading back goodput transparently guarantees up-to-date ingestion, existing user workloads will continue to work out-of-the-box without any migration steps.

Let me know what you think.

**Review by dipannita08 — CHANGES_REQUESTED** [review]

**Comment by findmyway** [comment]
@dipannita08 pushed `f564464` to address both comments. tl;dr:

**1. Public flush on `GoodputRecorder`** — done. Added `GoodputRecorder.flush()` that delegates to `self._cloud_logger.flush()` (no-op when `logging_enabled=False`, matching the existing `record_*` pattern). `_CloudLogger` stays internal.

**2. Encapsulate flushing in the read path** — went a different way after trying it. First implementation used a class-level `WeakSet` registry so a Calculator's read could find peer Recorder instances by `log_name` and flush their queues. After looking at it, it felt over-engineered for what it buys:

- The dominant production pattern is `GoodputMonitor` running in a subprocess (cross-process). No in-process registry helps there — staleness is bounded by `BackgroundThreadTransport.max_latency` (2 s default) plus the writer's atexit drain.
- The only case the registry would close is inline same-process `GoodputCalculator(...).get_*()` immediately after `record_*()`, where it would shave off up to 2 s of staleness. For goodput aggregates computed over minutes of training, 2 s is below measurement noise.
- The registry also had to handle stuck-writer corner cases (defensive `try/except` around peer flush) so a misbehaving writer can't block all readers — extra surface for an edge case.

Instead I documented the read-after-write window in `CHANGELOG.md` and the `read_cloud_logging_entries` docstring, and `Recorder.flush()` is there for the few cases that want a strict fence (unit tests, end-of-training finalization). Happy to add the registry back if you'd rather close the window strictly — it's ~10 lines.

**CHANGELOG** demoted from `### Changed (BREAKING)` to `### Changed`.

The MaxText TPU v6e test @YixuanWang-99 ran against the previous revision still applies — the write path didn't change, only the read-side semantics were documented and the public API surface shifted (no more `_CloudLogger.flush()` exposure).

File: ml_goodput_measurement/src/goodput.py [hunk 7]
```diff
@@ -102,19 +119,45 @@ def __init__(
...
+    )
+    self._async_handler = CloudLoggingHandler(
+        client=logging_client,
+        name=log_name,
+        transport=transport_factory,  # type: ignore[arg-type]
+    )
+    # One stdlib logger per log_name so multiple _CloudLogger instances
+    # don't share handlers; propagate=False keeps dict payloads off the
+    # root logger.
+    self._async_logger = logging.getLogger(f'_ml_goodput_async.{log_name}')
```
Comment (dipannita08) [inline ml_goodput_measurement/src/goodput.py]: Since `logging.getLogger` returns a global singleton, if the application potentially instantiates `_CloudLogger` multiple times in the same process (for example, running sequential jobs in a single notebook session or reusable worker process), handlers will accumulate on this shared logger singleton. When subsequent jobs write logs, both handlers will trigger, uploading every log entry twice to GCP.

We can prevent this handler leakage by adding id(self) to make the logger name unique per instance:

`self._async_logger = logging.getLogger(f'_ml_goodput_async.{log_name}.{id(self)}')`

Comment (findmyway) [reply ml_goodput_measurement/src/goodput.py]: Good catch — fixed in `ae72e7f`. Applied your suggested rename so each `_CloudLogger` owns a distinct stdlib logger:

```python
self._async_logger = logging.getLogger(
    f'_ml_goodput_async.{log_name}.{id(self)}'
)
```

Also dropped the now-redundant `if self._async_handler not in self._async_logger.handlers` guard — that check was actually load-bearing for the original bug to be silent in single-instance flows but never blocked re-attachment in the multi-instance case (each instance constructs a brand-new handler object, so `not in` was always true).

Added a regression test `test_two_instances_same_log_name_do_not_share_handlers` that constructs two `_CloudLogger`s with the same `log_name` and asserts (a) distinct backing loggers, (b) each logger holds only its own handler, (c) emitting on one doesn't fan out to the other. Verified the test fails on the pre-fix code with `AssertionError: unexpectedly identical: <Logger _ml_goodput_async.test-log (INFO)>` and passes after.

File: ml_goodput_measurement/src/goodput.py (hunk 1 above)
Comment (dipannita08) [reply ml_goodput_measurement/src/goodput.py]: LGTM.

**Review by dipannita08 — APPROVED** [review]
LGTM, thank you!

**Review by YixuanWang-99 — APPROVED** [review]
